docker/pusher.go

```diff
@@ -23,46 +29,66 @@ func NewPusher() *Pusher {
 }
 
 // Push the image of the Function.
-func (n *Pusher) Push(f bosonFunc.Function) (digest string, err error) {
-	// Check for the docker binary explicitly so that we can return
-	// an extra-friendly error message.
-	_, err = exec.LookPath("docker")
-	if err != nil {
-		err = errors.New("please install 'docker'")
-		return
-	}
+func (n *Pusher) Push(ctx context.Context, f bosonFunc.Function) (digest string, err error) {
 
 	if f.Image == "" {
 		return "", errors.New("Function has no associated image.  Has it been built?")
 	}
 
-	// set up the command, specifying a sanitized project name and connecting
-	// standard output and error.
-	cmd := exec.Command("docker", "push", f.Image)
+	parts := strings.Split(f.Image, "/")
+	if len(parts) < 1 {
+		return "", errors.Errorf("invalid image name: %q", f.Image)
+	}
+	registry := parts[0]
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create docker api client")
+	}
+
+	credentials, err := config.GetAllCredentials(nil)
```

Comment:
Not knowing the Docker API, how does this work? Does it just read your current `~/.docker/config.json`? What happens if a config doesn't exist?

Reply:
Fixed: it checks docker's and podman's config files. If the user is not logged in one of those two, then the user is prompted for credentials. Is that good @lance ?